File: groupon-copycat-frontend/src/app/models/customer.ts
/* tslint:disable:no-redundant-jsdoc */
import {Coupon} from './coupon';

export class Customer {
  private customerId: number;
  private password: string;
  private email: string;
  private firstName: string;
  private lastName: string;
  private coupons: Coupon[];

  constructor($customerId: number, $password: string, $email: string, $firstName: string, $lastName: string, $coupons: Coupon[]) {
    this.customerId = $customerId;
    this.password = $password;
    this.email = $email;
    this.firstName = $firstName;
    this.lastName = $lastName;
    this.coupons = $coupons;
  }

  /**
   * Getter $customerId
   * @return {number}
   */
  public get $customerId(): number {
    return this.customerId;
  }

  /**
   * Setter $customerId
   * @param {number} value
   */
  public set $customerId(value: number) {
    this.customerId = value;
  }

  /**
   * Getter $password
   * @return {string}
   */
  public get $password(): string {
    return this.password;
  }

  /**
   * Setter $password
   * @param {string} value
   */
  public set $password(value: string) {
    this.password = value;
  }

  /**
   * Getter $email
   * @return {string}
   */
  public get $email(): string {
    return this.email;
  }

  /**
   * Setter $email
   * @param {string} value
   */
  public set $email(value: string) {
    this.email = value;
  }

  /**
   * Getter $firstName
   * @return {string}
   */
  public get $firstName(): string {
    return this.firstName;
  }

  /**
   * Setter $firstName
   * @param {string} value
   */
  public set $firstName(value: string) {
    this.firstName = value;
  }

  /**
   * Getter $lastName
   * @return {string}
   */
  public get $lastName(): string {
    return this.lastName;
  }

  /**
   * Setter $lastName
   * @param {string} value
   */
  public set $lastName(value: string) {
    this.lastName = value;
  }

  /**
   * Getter $coupons
   * @return {Coupon[]}
   */
  public get $coupons(): Coupon[] {
    return this.coupons;
  }

  /**
   * Setter $coupons
   * @param {Coupon[]} value
   */
  public set $coupons(value: Coupon[]) {
    this.coupons = value;
  }

}

File: groupon-copycat/src/main/java/groupon_copycat/Spring_Boot/beans/Customer.java
package groupon_copycat.Spring_Boot.beans;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import javax.persistence.*;
import java.util.Set;

@Entity
@Table(name = "customers")
@JsonIgnoreProperties(ignoreUnknown = true)
public class Customer {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int customerId;
    @Column
    private String firstName, lastName;
    @Column(nullable = false, unique = true)
    private String email;
    @Column(nullable = false)
    private String password;
    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(name = "customers_coupons",
            joinColumns = @JoinColumn(name = "customer_id"),
            inverseJoinColumns = @JoinColumn(name = "coupon_id"))
    private Set<Coupon> coupons;

    public Customer() { // ctor for Hibernate
    }

    public Customer(String firstName, String lastName, String email, String password) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.email = email;
        this.password = password;
    }

    public void buyCoupon(Coupon coupon) {
        coupons.add(coupon);
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        this.lastName = lastName;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public String getPassword() {
        return password;
    }

    public void setPassword(String password) {
        this.password = password;
    }

    public int getCustomerId() {
        return customerId;
    }

    @Override
    public String toString() {
        return "Customer [customerId=" + customerId + ", firstName=" + firstName + ", lastName=" + lastName + ", email="
                + email + ", password=" + password + "]";
    }

    public Set<Coupon> getCoupons() {
        return coupons;
    }

}
